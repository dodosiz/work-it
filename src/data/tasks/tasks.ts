import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { v4 } from "uuid";
import { AppState } from "../store";
import { User } from "../users/users";

export interface TasksState {
	tasks: Task[];
	taskFormOpened: boolean;
}

export interface Task {
	id: string;
	title: string;
	description: string;
	assignee: User | undefined;
	dateFinished: string | undefined;
}

interface CreateTaskPayload {
	title: string;
	description: string;
	assignee: User | undefined;
}

interface FinishTaskPayload {
	taskId: string;
}

const initialState: TasksState = {
	tasks: [],
	taskFormOpened: false,
};

const tasksSlice = createSlice({
	name: "tasks",
	initialState,
	reducers: {
		createTask: (state, action: PayloadAction<CreateTaskPayload>) => {
			state.tasks.push({
				id: v4(),
				dateFinished: undefined,
				...action.payload,
			});
		},
		finishTask: (state, action: PayloadAction<FinishTaskPayload>) => {
			const task = state.tasks.find(
				(task) => task.id === action.payload.taskId
			);
			if (task) {
				task.dateFinished = new Date().toLocaleDateString();
			}
		},
		openTaskForm: (state) => {
			state.taskFormOpened = true;
		},
		closeTaskForm: (state) => {
			state.taskFormOpened = false;
		},
	},
});

export const { createTask, openTaskForm, closeTaskForm, finishTask } =
	tasksSlice.actions;
export const tasksReducer = tasksSlice.reducer;

export const toDoTasksSelector = (state: AppState) =>
	state.tasksState.tasks.filter((task) => task.dateFinished === undefined);
export const finishedTasksSelector = (state: AppState) =>
	state.tasksState.tasks.filter((task) => task.dateFinished !== undefined);
export const taskFormOpenedSelector = (state: AppState) =>
	state.tasksState.taskFormOpened;
