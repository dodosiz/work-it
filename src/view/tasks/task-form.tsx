import * as React from "react";
import { Button, Form } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { addNotification } from "../../data/notifications/notifications";
import {
	clickedTaskSelector,
	closeTaskForm,
	createTask,
	taskFormModeSelector,
} from "../../data/tasks/tasks";
import { usersSelector } from "../../data/users/users";

export function TaskForm() {
	const dispatch = useDispatch();
	const users = useSelector(usersSelector);
	const clickedTask = useSelector(clickedTaskSelector);
	const readOnly = useSelector(taskFormModeSelector) === "readonly";
	const [title, setTitle] = React.useState(
		clickedTask ? clickedTask.title : ""
	);
	const [description, setDescription] = React.useState(
		clickedTask ? clickedTask?.description : ""
	);
	const [userId, setUserId] = React.useState<string | undefined>(
		clickedTask ? clickedTask.assignee?.id : undefined
	);
	const [dirty, setDirty] = React.useState(false);
	const handleTitleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setTitle(event.target.value);
		setDirty(true);
	};
	const handleDescriptionChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setDescription(event.target.value);
		setDirty(true);
	};
	const handleAssigneeChange: React.ChangeEventHandler<HTMLSelectElement> = (
		event
	) => {
		setUserId(event.target.value);
		setDirty(true);
	};
	const handleSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		if (!readOnly) {
			const assignee = users.find((user) => user.id === userId);
			dispatch(createTask({ title, description, assignee }));
			dispatch(
				addNotification({
					message: `Created task: ${title}`,
				})
			);
			dispatch(closeTaskForm());
		}
	};
	const handleCancel = () => {
		dispatch(closeTaskForm());
	};
	const submitDisabled = () => {
		return !isValid() || !dirty;
	};
	const isValid = () => {
		return title && userId;
	};
	return (
		<Form onSubmit={handleSubmit}>
			<Form.Group className="mb-3" controlId="title">
				<Form.Label>Title:</Form.Label>
				<Form.Control
					readOnly={readOnly}
					type="text"
					value={title}
					onChange={handleTitleChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="description">
				<Form.Label>Description:</Form.Label>
				<Form.Control
					type="textarea"
					readOnly={readOnly}
					value={description}
					onChange={handleDescriptionChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="assignee">
				<Form.Select
					aria-label="Assignee"
					onChange={handleAssigneeChange}
					disabled={readOnly}
					value={userId}
					data-testid="assignee"
				>
					<option></option>
					{users.map((user) => (
						<option key={user.id} value={user.id}>
							{user.firstName + " " + user.lastName}
						</option>
					))}
				</Form.Select>
			</Form.Group>
			{!readOnly && (
				<>
					<Button
						className="me-2"
						disabled={submitDisabled()}
						variant="primary"
						type="submit"
					>
						Save
					</Button>
					<Button variant="secondary" onClick={() => handleCancel()}>
						Cancel
					</Button>
				</>
			)}
		</Form>
	);
}
