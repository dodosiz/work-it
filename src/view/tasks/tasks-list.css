.task {
	box-shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;
	margin: 2em 0;
}
