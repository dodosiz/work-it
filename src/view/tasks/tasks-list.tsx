import * as React from "react";
import { Button, Card, Col, Row } from "react-bootstrap";
import {
	BsCheckCircle,
	BsCheckCircleFill,
	BsFillTrashFill,
	BsPencilFill,
} from "react-icons/bs";
import { useDispatch, useSelector } from "react-redux";
import { addNotification } from "../../data/notifications/notifications";
import {
	deleteTask,
	finishTask,
	openTaskForm,
	Task,
	tasksSelector,
} from "../../data/tasks/tasks";
import "./tasks-list.css";

export function TasksList(props: { mode: "todo" | "done" }) {
	const tasks = useSelector(tasksSelector(props.mode));
	if (tasks.length === 0) {
		return (
			<p>
				{props.mode === "todo" ? "No tasks found." : "No finished tasks found."}
			</p>
		);
	}
	return (
		<>
			{tasks.sort(sortByDate).map((task) => (
				<TaskCard key={task.id} task={task} mode={props.mode} />
			))}
		</>
	);
}

function sortByDate(task1: Task, task2: Task) {
	if (task1.dateFinished && task2.dateFinished) {
		return task2.dateFinished.localeCompare(task1.dateFinished);
	} else {
		return 0;
	}
}

interface TaskCardProps {
	mode: "todo" | "done";
	task: Task;
}

function TaskCard(props: TaskCardProps) {
	const fadeOutTimeout = 1500; // a little faster then the CSS transition
	const [checked, setChecked] = React.useState(false);
	const dispatch = useDispatch();
	const handleCheck = () => {
		setChecked(true);
		setTimeout(() => {
			dispatch(finishTask({ taskId: props.task.id }));
		}, fadeOutTimeout);
	};
	const handleEdit = () => {
		dispatch(
			openTaskForm({
				taskFormMode: "edit",
				clickedTask: props.task,
			})
		);
	};
	const handleDelete = () => {
		dispatch(deleteTask({ taskId: props.task.id }));
		dispatch(
			addNotification({
				message: `Deleted task "${props.task.title}".`,
			})
		);
	};
	const handleTitleClick = () => {
		dispatch(
			openTaskForm({
				taskFormMode: "readonly",
				clickedTask: props.task,
			})
		);
	};
	return (
		<Card className="task" style={checked ? { opacity: "0%" } : {}}>
			<Card.Body>
				<Row>
					<Col md={1}>
						<CheckBox
							taskId={props.task.id}
							checked={props.mode === "todo" ? checked : true}
							handleCheck={props.mode === "todo" ? handleCheck : undefined}
						/>
					</Col>
					<Col md={5}>
						<span
							className="pointer"
							data-testid={`title-${props.task.id}`}
							onClick={handleTitleClick}
						>
							{props.task.title}
						</span>
					</Col>
					<Col md={4}>
						{props.task.assignee?.firstName +
							" " +
							props.task.assignee?.lastName}
					</Col>
					{props.mode === "done" && (
						<Col md={2}>{`Finished on: ${props.task.dateFinished}`}</Col>
					)}
					{props.mode === "todo" && (
						<Col md={2}>
							<Button
								data-testid={`edit-button-${props.task.id}`}
								variant="outline-primary"
								className="me-2"
								onClick={handleEdit}
							>
								<BsPencilFill />
							</Button>
							<Button
								data-testid={`delete-button-${props.task.id}`}
								variant="outline-danger"
								onClick={handleDelete}
							>
								<BsFillTrashFill />
							</Button>
						</Col>
					)}
				</Row>
			</Card.Body>
		</Card>
	);
}

interface CheckBoxProps {
	taskId: string;
	checked: boolean;
	handleCheck?: () => void;
}

function CheckBox(props: CheckBoxProps) {
	return (
		<div
			className={`${props.checked ? "done" : "undone"}`}
			data-testid={`check-${props.taskId}`}
		>
			{!props.checked && (
				<BsCheckCircle
					className="pointer"
					onClick={props.handleCheck}
					size={"1.5em"}
					data-testid={`check-box-${props.taskId}`}
				/>
			)}
			{props.checked && <BsCheckCircleFill size={"1.5em"} />}
		</div>
	);
}
