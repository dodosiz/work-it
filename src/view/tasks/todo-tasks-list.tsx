import * as React from "react";
import { Card, Col, Row } from "react-bootstrap";
import { Task } from "../../data/tasks/tasks";
import "./tasks-list.css";

export function TodoTasksList(props: { tasks: Task[] }) {
	if (props.tasks.length === 0) {
		return <p>No tasks to do. Start by creating a new task.</p>;
	}
	return (
		<>
			{props.tasks.map((task) => (
				<TodoTask key={task.id} task={task} />
			))}
		</>
	);
}

function TodoTask(props: { task: Task }) {
	return (
		<Card className="task">
			<Card.Body>
				<Row>
					<Col>{props.task.title}</Col>
					<Col>
						{props.task.assignee?.firstName +
							" " +
							props.task.assignee?.lastName}
					</Col>
				</Row>
			</Card.Body>
		</Card>
	);
}
