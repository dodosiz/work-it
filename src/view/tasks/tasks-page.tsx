import * as React from "react";
import { Container, Navbar, Nav, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import {
	closeTaskForm,
	openTaskForm,
	taskFormOpenedSelector,
	toDoTasksSelector,
} from "../../data/tasks/tasks";
import { ModalDialog } from "../modal/modal-dialog";
import { TaskForm } from "./task-form";
import { TodoTasksList } from "./todo-tasks-list";

export function TasksPage() {
	const dispatch = useDispatch();
	const tasks = useSelector(toDoTasksSelector);
	const taskFormOpened = useSelector(taskFormOpenedSelector);
	const handleCloseTaskForm = () => {
		dispatch(closeTaskForm());
	};
	const handleCreateTaskClick = () => {
		dispatch(openTaskForm());
	};
	return (
		<>
			<ModalDialog
				handleClose={handleCloseTaskForm}
				show={taskFormOpened}
				title="Create new task"
			>
				<TaskForm />
			</ModalDialog>
			<Navbar expand="lg" variant="light" bg="light">
				<Container>
					<Nav className="me-auto">
						<Button variant="success" onClick={handleCreateTaskClick}>
							Create Task
						</Button>
					</Nav>
				</Container>
			</Navbar>
			<Container>
				<TodoTasksList tasks={tasks} />
			</Container>
		</>
	);
}
