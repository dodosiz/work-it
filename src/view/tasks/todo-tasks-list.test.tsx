import * as React from "react";
import { Task } from "../../data/tasks/tasks";
import { TodoTasksList } from "./todo-tasks-list";
import { render } from "@testing-library/react";

describe("Todo tasks list:", () => {
	test("Snapshot test", () => {
		const tasks: Task[] = [
			{
				id: "task1",
				assignee: {
					id: "user1",
					firstName: "Alejandro",
					lastName: "Matthews",
					role: "Project manager",
				},
				title: "Plan the requirements for the next release.",
				dateFinished: undefined,
				description: "",
			},
			{
				id: "task2",
				assignee: {
					id: "user1",
					firstName: "Alejandro",
					lastName: "Matthews",
					role: "Project manager",
				},
				title: "Meet with the stakeholders to discuss the new requirements.",
				dateFinished: undefined,
				description: "",
			},
		];
		const component = render(<TodoTasksList tasks={tasks} />);
		expect(component.container).toMatchSnapshot();
	});
});
