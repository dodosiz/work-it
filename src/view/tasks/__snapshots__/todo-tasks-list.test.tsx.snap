// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo tasks list: Snapshot test 1`] = `
<div>
  <div
    class="task card"
  >
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          Plan the requirements for the next release.
        </div>
        <div
          class="col"
        >
          Alejandro Matthews
        </div>
      </div>
    </div>
  </div>
  <div
    class="task card"
  >
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          Meet with the stakeholders to discuss the new requirements.
        </div>
        <div
          class="col"
        >
          Alejandro Matthews
        </div>
      </div>
    </div>
  </div>
</div>
`;
